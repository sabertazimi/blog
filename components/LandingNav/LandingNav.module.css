.nav {
  @apply fixed left-0 top-0 z-20 flex h-full w-full flex-col items-center justify-center;
}

.banner {
  @apply fixed left-0 top-0 z-10 block h-full w-full bg-black/50;
  @apply backdrop-brightness-110 backdrop-contrast-125 backdrop-saturate-150;
}

.icon {
  @apply fixed left-12 top-12 z-100 cursor-pointer border-none bg-transparent text-lg font-extrabold text-light md:text-4xl;
}
