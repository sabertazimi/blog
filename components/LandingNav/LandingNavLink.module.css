.link {
  @apply relative mx-6 my-0 block cursor-pointer px-0 pt-0 pb-3;
}

.link a {
  @apply text-4xl font-extrabold leading-none text-light md:text-7xl;
}

.link::after {
  @apply absolute left-0 top-full h-2 w-full;
  @apply bg-gradient-primary rounded-lg;

  content: '';
}

.tooltip {
  @apply mb-16;
}
