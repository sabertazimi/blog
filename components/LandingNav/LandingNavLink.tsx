import type { ReactNode } from 'react'
import Link from '@/components/Link'
import { Bounce } from '@/components/Motion'
import Tooltip from '@/components/Tooltip'
import styles from './LandingNavLink.module.css'

interface Props {
  title: string
  href: string
  children: ReactNode
}

function LandingNavLink({ title, href, children }: Props): JSX.Element {
  return (
    <Tooltip className={styles.tooltip} placement="left" title={title}>
      <Bounce>
        <span className={styles.link} role="link">
          <Link href={href}>{children}</Link>
        </span>
      </Bounce>
    </Tooltip>
  )
}

export default LandingNavLink
