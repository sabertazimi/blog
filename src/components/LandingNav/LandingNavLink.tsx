import { Tooltip } from 'antd';
import classNames from 'classnames';
import { Link } from 'gatsby';
import type { ReactNode } from 'react';
import React from 'react';

interface Props {
  title: string;
  to: string;
  children: ReactNode;
}

const LandingNavLink = ({ title, to, children }: Props): JSX.Element => (
  <Tooltip className="mb-16" placement="left" title={title}>
    <Link to={to} role="link">
      <span
        className={classNames(
          'block px-0 pt-0 pb-3 mx-6 my-0 border-solid border-b-6 border-light',
          'text-4xl font-extrabold leading-none cursor-pointer text-light md:text-7xl',
          'transition-all transform-gpu scale-100 opacity-80 hover:opacity-100 hover:scale-110'
        )}
      >
        {children}
      </span>
    </Link>
  </Tooltip>
);

export default LandingNavLink;
