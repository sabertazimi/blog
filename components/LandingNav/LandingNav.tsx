'use client'

import type { Route } from '@/types'
import type { MotionProps } from 'framer-motion'
import type { KeyboardEvent } from 'react'
import { Hamburger } from '@/components/Icons'
import { Switch } from '@/components/Motion'
import { routes as defaultRoutes } from '@/config'
import { useCallback, useState } from 'react'
import styles from './LandingNav.module.css'
import LandingNavLink from './LandingNavLink'

interface Props {
  routes?: Route[]
}

const navVariants: MotionProps['variants'] = {
  open: {
    x: 0,
    opacity: 1,
  },
  close: {
    x: '-100%',
    opacity: 0,
  },
}

const navTransition: MotionProps['transition'] = {
  type: 'tween',
  ease: 'easeInOut',
  duration: 0.4,
}

const bannerVariants: MotionProps['variants'] = {
  open: {
    x: 0,
  },
  close: {
    x: '-100%',
  },
}

const bannerTransition: MotionProps['transition'] = {
  type: 'tween',
  ease: 'circInOut',
  duration: 0.4,
}

function LandingNav({ routes = defaultRoutes }: Props): JSX.Element {
  const [expanded, setExpanded] = useState(false)

  const handleClick = useCallback(() => {
    setExpanded(expanded => !expanded)
  }, [])

  const handleKeyDown = useCallback((event: KeyboardEvent<HTMLSpanElement>) => {
    if (event.key === 'Enter')
      setExpanded(expanded => !expanded)
  }, [])

  return (
    <>
      <Hamburger
        onClick={handleClick}
        onKeyDown={handleKeyDown}
        className={styles.icon}
        data-testid="hamburger-button"
      />
      <Switch
        className={styles.nav}
        role="navigation"
        open={expanded}
        variants={navVariants}
        transition={navTransition}
      >
        {routes.map(route => (
          <LandingNavLink key={route.id} title={route.title} href={route.path}>
            {route.name}
          </LandingNavLink>
        ))}
      </Switch>
      <Switch
        className={styles.banner}
        role="banner"
        open={expanded}
        variants={bannerVariants}
        transition={bannerTransition}
      />
    </>
  )
}

export default LandingNav
