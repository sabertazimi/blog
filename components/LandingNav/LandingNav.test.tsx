import { act, render, screen } from '@testing-library/react';
import LandingNav from './LandingNav';

describe('LandingNav', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  test('should render routes correctly (snapshot)', () => {
    const { container } = render(<LandingNav />);

    act(() => {
      jest.advanceTimersByTime(1000);
    });

    expect(container).toMatchSnapshot();
  });

  test('should render route with correct structure', () => {
    render(<LandingNav />);

    const nav = screen.getByRole('navigation');
    const icon = screen.getByRole('img');

    expect(nav).toBeInTheDocument();
    expect(icon).toBeInTheDocument();

    const navButton = screen.getByTestId('hamburger-button');

    expect(navButton).toBeInTheDocument();
    expect(navButton).toContainElement(icon);

    const navLinks = screen.getAllByRole('link');

    navLinks.forEach(navLink => expect(navLink).toBeInTheDocument());
    navLinks.forEach(navLink => expect(nav).toContainElement(navLink));
  });
});
