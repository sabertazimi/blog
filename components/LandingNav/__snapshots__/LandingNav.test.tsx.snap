// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandingNav should render routes correctly (snapshot) 1`] = `
<div>
  <span
    aria-label="hamburger"
    class="anticon icon"
    data-testid="hamburger-button"
    role="img"
    tabindex="-1"
  >
    <svg
      data-testid="hamburger-icon"
      fill="currentColor"
      height="2em"
      p-id="7522"
      tabindex="0"
      viewBox="0 0 1024 1024"
      width="2em"
    >
      <path
        d="M904 160H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
        style="transform: none; transform-origin: 0px 0px;"
      />
      <path
        d="M904 472H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
        style="transform: none; transform-origin: 0px 0px;"
      />
      <path
        d="M904 784H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
        style="transform: none; transform-origin: 0px 0px;"
      />
    </svg>
  </span>
  <div
    class="nav"
    role="navigation"
    style="opacity: 0; transform: translateX(-100%) translateZ(0);"
  >
    <div
      class="inline-flex tooltip"
      style="opacity: 0; transform: translateX(20px) translateZ(0);"
      tabindex="0"
    >
      <span
        class="link"
        role="link"
      >
        <a
          href="/posts"
        >
          Posts
        </a>
      </span>
    </div>
    <div
      class="inline-flex tooltip"
      style="opacity: 0; transform: translateX(20px) translateZ(0);"
      tabindex="0"
    >
      <span
        class="link"
        role="link"
      >
        <a
          href="/tags"
        >
          Tags
        </a>
      </span>
    </div>
    <div
      class="inline-flex tooltip"
      style="opacity: 0; transform: translateX(20px) translateZ(0);"
      tabindex="0"
    >
      <span
        class="link"
        role="link"
      >
        <a
          href="/books"
        >
          Books
        </a>
      </span>
    </div>
    <div
      class="inline-flex tooltip"
      style="opacity: 0; transform: translateX(20px) translateZ(0);"
      tabindex="0"
    >
      <span
        class="link"
        role="link"
      >
        <a
          href="/about"
        >
          About
        </a>
      </span>
    </div>
  </div>
  <div
    class="banner"
    role="banner"
    style="transform: translateX(-100%) translateZ(0);"
  />
</div>
`;
